INT-4162: TCP/UDP DSL

garyrussell opened the pull request:
JIRA: https://jira.spring.io/browse/INT-4162

INT-4162: IP DSL - Phase I Connection Factories

Phase 2 - Adapters/Gateways

Phase 3 - UDP

garyrussell committed:
INT-4162: TCP/UDP DSL

JIRA: https://jira.spring.io/browse/INT-4162

INT-4162: IP DSL - Phase I Connection Factories

Phase 2 - Adapters/Gateways

Phase 3 - UDP

artembilan commented on spring-integration-ip/src/main/java/org/springframework/integration/ip/dsl/TcpInboundChannelAdapterSpec.java:
```diff
@@ -0,0 +1,69 @@
+ * @since 5.0
+ *
+ */
+public class TcpInboundChannelAdapterSpec<S extends TcpInboundChannelAdapterSpec<S>>
```
Well, I see that this generic variant works well in the target code.
So, now I'm not sure if we should restrict with target type without generic modification or leave it with that and let to extend in the future easy...
But, eh., typically we don't do that.
And I'm not sure that we are going to extend `TcpReceivingChannelAdapter` class to have some `TcpInboundChannelAdapterSpec` extension as well.

garyrussell committed:
Polishing - PR Comments

Remove unnecessary generics.

garyrussell commented:
@artembilan One issue with this implementation is...

    ...from(Tcp.inbouncChannelAdapter(Tcp.server(1234).get())...

The connection factory is not a bean, so in the tests I used

    ...from(Tcp.inboundGateway(server1()))...

so the connection factory is initialized.

Any suggestions for making the first one work - if we inject a Spec instead of the result of the `get()` perhaps ??

artembilan commented:
The `TcpInboundChannelAdapterSpec` can populate provided `ConnectionFactory` to the `ComponentsRegistration` and DSL `BPP` will take care if it need to register a bean or such one is already there.

artembilan commented on spring-integration-ip/src/main/java/org/springframework/integration/ip/dsl/AbstractConnectionFactorySpec.java:
```diff
@@ -0,0 +1,223 @@
+	}
+
+	/**
+	 * @param soKeepAlive the socket keepalive option.
```
Looks like a typo :smile: : keep alive ?

artembilan commented on spring-integration-ip/src/main/java/org/springframework/integration/ip/dsl/AbstractConnectionFactorySpec.java:
```diff
@@ -0,0 +1,223 @@
+	}
+
+	/**
+	 * @param taskExecutor the task excecutor.
```
typo: ex~~c~~ecutor

artembilan commented on spring-integration-ip/src/main/java/org/springframework/integration/ip/dsl/TcpClientConnectionFactorySpec.java:
```diff
@@ -0,0 +1,42 @@
+import org.springframework.integration.ip.tcp.connection.TcpNioClientConnectionFactory;
+
+/**
+ * An {@link IntegrationComponentSpec} for {@link AbstractClientConnectionFactory}s.
```
The JavaDocs doesn't reflect reality of this class.

artembilan commented on spring-integration-ip/src/main/java/org/springframework/integration/ip/dsl/TcpServerConnectionFactorySpec.java:
```diff
@@ -0,0 +1,62 @@
+import org.springframework.integration.ip.tcp.connection.TcpNioServerConnectionFactory;
+
+/**
+ * An {@link IntegrationComponentSpec} for {@link AbstractServerConnectionFactory}s.
```
The same JavaDocs problem here.

artembilan commented on spring-integration-ip/src/main/java/org/springframework/integration/ip/dsl/TcpOutboundGatewaySpec.java:
```diff
@@ -0,0 +1,42 @@
+		this.target.setConnectionFactory(connectionFactory);
+	}
+
+	public TcpOutboundGatewaySpec remoteTimeout(long remoteTimeout) {
```
Would you mind to elaborate why this hasn't been addressed?
Thanks

artembilan commented on spring-integration-ip/src/main/java/org/springframework/integration/ip/dsl/Udp.java:
```diff
@@ -0,0 +1,48 @@
+	}
+
+	public static UdpOutboundChannelAdapterSpec outboundMulticastAdapter(String destinationExpression) {
+		return new UdpMulticastOutboundChannelAdapterSpec(destinationExpression);
```
I think the return type of the method must be `UdpMulticastOutboundChannelAdapterSpec` as well.
Otherwise we don't have access to the `timeToLive()`  method from the target code.

artembilan commented on spring-integration-ip/src/main/java/org/springframework/integration/ip/dsl/Udp.java:
```diff
@@ -0,0 +1,48 @@
+		return new UdpInboundChannelAdapterSpec(port, multicastGroup);
+	}
+
+	public static UdpOutboundChannelAdapterSpec outboundAdapter(String destinationExpression) {
```
Why don't we have `UnicastSendingMessageHandler(String host, int port)` variant factory?

Plus it looks like we can have `Expression` and `Function` variants as well.

garyrussell replied:
>why don't - I don't like all those ctors and the URL version (added recently) seems simple enough to me.

I am not sure what you mean by `Function` and `Expression` variants - `Expression` is painful for DSL use; please explain the benefit of a `Function` version.

artembilan replied:
`Expression` variant can be `LiteralExpression`, `ValueExpression` or any other impl. That is much convenient than just have only `String` variant.

The `Function` case would be like this:
```
Udp.outboundAdapter(m -> m.getHeaders().get("udp_dest"))
```
which you should find as more convenient for Java than SpEL.
The impl can look like this:
```
public static UdpOutboundChannelAdapterSpec outboundAdapter(Function<Message<?>, Object> destinationFunction) {
	return new UdpOutboundChannelAdapterSpec(new FunctionExpression<>(destinationFunction));
}
```

garyrussell replied:
Never mind - I see it now - missed your earlier comment about remoteTimeout.

garyrussell replied:
Yeah - I've added the Function variant - still not so sure about `Expression` - I don't see that variant in the Jms specs for destinations, for example.

artembilan replied:
Maybe that was just an omission?
How about `SupplierExpression`? 
Does it have some power on you for conviction?  :smile:

artembilan commented on spring-integration-ip/src/main/java/org/springframework/integration/ip/dsl/UdpInboundChannelAdapterSpec.java:
```diff
@@ -0,0 +1,155 @@
+import org.springframework.scheduling.TaskScheduler;
+
+/**
+ * A {@link MessageProducerSpec} for {@link UnicastReceivingChannelAdapter}UdpInboundChannelAdapterSpec.
```
Typo. The `UdpInboundChannelAdapterSpec` in the end looks odd...

artembilan commented on spring-integration-ip/src/main/java/org/springframework/integration/ip/dsl/UdpInboundChannelAdapterSpec.java:
```diff
@@ -0,0 +1,155 @@
+	}
+
+	/**
+	 * @param soSendBufferSize set the send bufffer size socket option.
```
Typo: triple buffer for `f` :smile:

artembilan commented on spring-integration-ip/src/main/java/org/springframework/integration/ip/dsl/UdpOutboundChannelAdapterSpec.java:
```diff
@@ -0,0 +1,121 @@
+	 * @return the spec.
+	 * @see UnicastSendingMessageHandler#setSocketExpression(Expression)
+	 */
+	public UdpOutboundChannelAdapterSpec socketExpression(Expression socketExpression) {
```
`socketFunction()` with wrapping to the `FunctionExpression` would be good, too.

artembilan commented on spring-integration-ip/src/main/java/org/springframework/integration/ip/udp/MulticastSendingMessageHandler.java:
```diff
@@ -99,6 +102,28 @@ public MulticastSendingMessageHandler(String address, int port,
 	}
 
+	/**
+	 * Construct UnicastSendingMessageHandler based on the destination SpEL expression to
```
s/UnicastSendingMessageHandler/MulticastSendingMessageHandler

artembilan commented on spring-integration-ip/src/main/java/org/springframework/integration/ip/udp/MulticastSendingMessageHandler.java:
```diff
@@ -99,6 +102,28 @@ public MulticastSendingMessageHandler(String address, int port,
+	}
+
+	/**
+	 * Construct UnicastSendingMessageHandler based on the destination SpEL expression to
```
DITTO

artembilan commented on spring-integration-ip/src/test/java/org/springframework/integration/ip/dsl/IpIntegrationTests.java:
```diff
@@ -0,0 +1,178 @@
+				.handle(Tcp.outboundGateway(this.client1))
+				.transform(Transformers.objectToString());
+		IntegrationFlowRegistration theFlow = this.flowContext.registration(flow).register();
+		assertThat(theFlow.getMessagingTemplate().convertSendAndReceive("foo", String.class), equalTo("FOO"));
```
Cool observation for unit testing :smile:

artembilan reviewed (changes requested):
I don't see `ComponentsRegistration` implementation.
Or you decided to stay away from that (at least for now) or have just missed my comment as I don't see email notification for your latest commit...

garyrussell committed:
More Polishing - PR Comments

artembilan commented:
```
:spring-integration-ip:checkstyleMain[ant:checkstyle] [ERROR] /home/travis/build/spring-projects/spring-integration/spring-integration-ip/src/main/java/org/springframework/integration/ip/dsl/UdpOutboundChannelAdapterSpec.java:21:8: Unused import - org.springframework.expression.Expression. [UnusedImports]
```
:smile:

garyrussell committed:
Checkstyle/Javadoc Fixes

artembilan commented on spring-integration-ip/src/main/java/org/springframework/integration/ip/dsl/UdpOutboundChannelAdapterSpec.java:
```diff
@@ -0,0 +1,124 @@
+	 * @return the spec.
+	 * @see UnicastSendingMessageHandler#setSocketExpression(org.springframework.expression.Expression)
+	 */
+	public UdpOutboundChannelAdapterSpec socketExpression(Function<Message<?>, ?> socketExpressionFunction) {
```
It's definitely `socketFunction()`.
The fact transformation to a `FunctionExpression` is out of end-user API.

Plus this must be like `Function<Message<?>, DatagramSocket>`, just because:
```
if (this.socketExpression != null) {
	socket = this.socketExpression.getValue(this.evaluationContext, message, DatagramSocket.class);
}
```
therefore we restrict the target code with particular type and that is much easier to write the final lambda.

artembilan commented on spring-integration-ip/src/main/java/org/springframework/integration/ip/dsl/UdpOutboundChannelAdapterSpec.java:
```diff
@@ -0,0 +1,124 @@
+		super();
+	}
+
+	UdpOutboundChannelAdapterSpec(String destinationExpression) {
```
Since there is `UnicastSendingMessageHandler(Expression destinationExpression)`, I would like to see `UdpOutboundChannelAdapterSpec(Function<Message<?>, Object> destinationFunction)`

artembilan commented on spring-integration-ip/src/main/java/org/springframework/integration/ip/dsl/TcpInboundChannelAdapterSpec.java:
```diff
@@ -0,0 +1,82 @@
+
+	TcpInboundChannelAdapterSpec(AbstractConnectionFactory connectionFactory) {
+		super(new TcpReceivingChannelAdapter());
+		this.connectionFactory = connectionFactory;
```
Well, I see your point about `ComponentsRegistration`. 
How about to have alternative ctor for the `AbstractConnectionFactorySpec` and populate for registration just only result of that?
Any `AbstractConnectionFactory` variant should be already supplied as a `@Bean`
WDYT?

I suggests this because I don't know how to answer to your concern in the https://github.com/spring-projects/spring-integration-java-dsl/issues/137 :smile:

garyrussell committed:
ComponentsRegistration Improvements

Add ...Spec CTORs to avoid the issue described here:
https://github.com/spring-projects/spring-integration-java-dsl/issues/137

Also other PR comments.

artembilan commented on spring-integration-ip/src/main/java/org/springframework/integration/ip/dsl/UdpOutboundChannelAdapterSpec.java:
```diff
@@ -102,12 +107,12 @@ public UdpOutboundChannelAdapterSpec ackCounter(int ackCounter) {
 	 */
-	public UdpOutboundChannelAdapterSpec socketExpression(Function<Message<?>, ?> socketExpressionFunction) {
-		this.target.setSocketExpression(new FunctionExpression<>(socketExpressionFunction));
+	public UdpOutboundChannelAdapterSpec socketExpression(Function<Message<?>, DatagramSocket> socketFunction) {
```
Do you agree with me that `socketFunction()` sounds better as a name for this kind of method ?

garyrussell replied:
>We spent already enough time on this PR

Thanks for your patience - this was my first complete module; hopefully my next commit will be the last.

artembilan commented on spring-integration-ip/src/main/java/org/springframework/integration/ip/dsl/UdpOutboundChannelAdapterSpec.java:
```diff
@@ -0,0 +1,129 @@
+		this.target = new UnicastSendingMessageHandler(destinationExpression);
+	}
+
+	UdpOutboundChannelAdapterSpec(Function<Message<?>, ?> destinationFunction) {
```
This ctor isn't covered in the `Udp`.
The same concern for the `UdpMulticastOutboundChannelAdapterSpec`.
Also I see a generic problem:
```
Udp.outboundMulticastAdapter("headers['udp_dest']")
					.lengthCheck(true)
					.timeToLive(10)
```
We can't reach an inheritors methods if we call the super.
So, this `UdpOutboundChannelAdapterSpec` must be with generic specification for itself. Exactly what you had before on the `TcpInboundChannelAdapterSpec` for example.
Or what we have on the `JmsOutboundChannelAdapterSpec` or `AbstractConnectionFactorySpec`.

garyrussell committed:
Fix UDP Spec Inheritance

artembilan commented:
The Travis failure isn't related to your work, but I wonder if you wouldn't mind taking a fresh look to that `JdbcLockRegistryLeaderInitiatorTests`.
Thanks

artembilan commented on spring-integration-ip/src/main/java/org/springframework/integration/ip/dsl/AbstractUdpOutboundChannelAdapterSpec.java:
```diff
@@ -25,24 +25,25 @@
 
 /**
- * A {@link MessageHandlerSpec} for {@link UnicastSendingMessageHandler}s.
+ * A {@link MessageHandlerSpec} for UDB {@link org.springframework.messaging.MessageHandler}s.
```
Typo: UDP ?

artembilan commented on spring-integration-ip/src/main/java/org/springframework/integration/ip/dsl/AbstractUdpOutboundChannelAdapterSpec.java:
```diff
@@ -25,24 +25,25 @@
- * A {@link MessageHandlerSpec} for {@link UnicastSendingMessageHandler}s.
+ * A {@link MessageHandlerSpec} for UDB {@link org.springframework.messaging.MessageHandler}s.
  *
+ * @param <S> the target {@link UdpUnicastOutboundChannelAdapterSpec} implementation type.
```
s/UdpUnicastOutboundChannelAdapterSpec/AbstractUdpOutboundChannelAdapterSpec

garyrussell committed:
Jdbc Lock Test Log Adjuster

garyrussell commented:
It's too hard to debug travis problems without test artifacts - added a log adjuster to get info if it occurs on Bamboo.

artembilan commented on spring-integration-ip/src/test/java/org/springframework/integration/ip/dsl/IpIntegrationTests.java:
```diff
@@ -128,6 +128,13 @@ public void testUdp() {
+		Udp.outboundMulticastAdapter("headers['udp_dest']")
+			.lengthCheck(true)
+			.timeToLive(10);
+	}
```
This test doesn't look finished: no one assertion.
I understand the premise of this, but looks like a visual observation and compiler approval make a staff.
OTOH I understand also that there might be a case when we change the API and just don't have the code to prove a support for proper method chain. 
Maybe it really would be good to finish the test for the valid `MulticastSendingMessageHandler` use-case?..

artembilan replied:
OK. I'll just add in the end an assertion  that we have proper `instanceOf` after `Spec` evaluation.

artembilan commented:
According the latest Travis report we have some problem in the `MongoDbInboundChannelAdapterIntegrationTests`.
I think there is a race condition with all those `fixed-rate` in config.
And our `testWithOnSuccessDisposition()` just received two message for the same document. Just because the poller is able to start a new polling task because the previous hasn't finished during expected `fixed-rate="200"`.

Right, that isn't task of your PR, but does it matter when and how do we fix it ?

Your thoughts ?
Thanks

garyrussell committed:
Fix Mongo Test Race Conditions

artembilan commented:
Merged as https://github.com/spring-projects/spring-integration/commit/65ec234697e3577b5b19e6d2aca6fc04db7d501c
